iree/turbine/kernel/wave/codegen/handlers.py

```diff
@@ -935,49 +935,75 @@ def handle_iterate(emitter: WaveEmitter, node: fx.Node):
+    flat_init_arg_types = [arg.ir_value.type for arg in flat_init_args]
+
     # Initialize while loop
     init_value = cast_py_value(emitter, start).ir_value
     if isinstance(init_value.type, VectorType):
         init_value = vector_d.extractelement(init_value, [0])
-    elif isinstance(init_value.type, IntegerType):
+    if isinstance(init_value.type, IntegerType):
         init_value = arith_d.index_cast(IndexType.get(), init_value)
 
     assert isinstance(
         init_value.type, IndexType
     ), f"Unhandled start type: {init_value.type}"
 
-    whileOp = scf_d.WhileOp([init_value.type], [init_value])
-    whileOp.before.blocks.append(*[init_value.type])
-    whileOp.after.blocks.append(*[init_value.type])
+    init_args = [arg.ir_value for arg in flat_init_args] + [init_value]
+    init_arg_types = flat_init_arg_types + [init_value.type]
+    whileOp = scf_d.WhileOp(init_arg_types, init_args)
+    whileOp.before.blocks.append(*init_arg_types)
+    whileOp.after.blocks.append(*init_arg_types)
 
     # Before block: condition.
-    current_value = whileOp.before.blocks[0].arguments[0]
+    current_values = whileOp.before.blocks[0].arguments
     with InsertionPoint(whileOp.before.blocks[0]):
         # Replace the axis with a temporary variable when generating the condition
         # to avoid conflicts with the actual value of the axis.
         condition = condition.subs({axis: index_symbol("$TMP")})
         subs = add_emitter_subs(emitter)
-        subs[index_symbol("$TMP")] = current_value
+        subs[index_symbol("$TMP")] = current_values[-1]
         condition = gen_sympy_index(subs, condition)
-        scf_d.ConditionOp(condition, [current_value])
+        scf_d.ConditionOp(condition, current_values)
 
     # After block: body.
-    current_value = whileOp.after.blocks[0].arguments[0]
+    current_values = whileOp.after.blocks[0].arguments
     with InsertionPoint(whileOp.after.blocks[0]):
         subgraph = emitter.trace.get_subgraph(subgraph)
+        # Map the captured variables from the root graph to the subgraph
         for root_v, subgraph_v in zip(
             implicit_capture, get_custom(node).captured_vars(subgraph)
         ):
             emitter._node_values[subgraph_v] = emitter.lookup_node_values(root_v)
-        emitter.induction_vars[axis] = current_value
-        emitter._emit_graph(subgraph)
-        scf_d.YieldOp([emitter.dynamic_dims[axis]])
+
+        # Map the iteration variable
+        emitter.induction_vars[axis] = current_values[-1]
+
+        # Map the iteration arguments
+        iter_args = get_custom(node).iter_args(subgraph)
+        for i, arg in enumerate(iter_args):
+            if i < len(current_values) - 1:
+                emitter.bind_node_proxy(arg, IRProxyValue(current_values[i]))
+
+        # Emit the subgraph
+        return_values = emitter._emit_graph(subgraph)
+        if all(x is None for x in return_values):
```

Comment:
Is this the case where there is no init_args aside from the condition variable? Can we add some comments/docs here?

Reply:
yeah this is just a pure while with no init args. Sure will add a comment.